src/main/java/top/seraphjack/simplelogin/server/storage/StorageProviderFile.java

```diff
@@ -106,16 +108,39 @@ public boolean dirty() {
         return dirty;
     }
 
+    @Override
+    public void setLastPosition(String username, Position pos) {
+        if (entries.containsKey(username)) {
+            POJOUserEntry entry = entries.get(username);
+            entry.posX = pos.getX();
+            entry.posY = pos.getY();
+            entry.posZ = pos.getZ();
+        }
+    }
+
+    @Override
+    public Position getLastPosition(String username) {
+        if (entries.containsKey(username)) {
+            POJOUserEntry entry = entries.get(username);
+            return new Position(entry.posX, entry.posY, entry.posZ);
+        }
+        return SLConstants.defaultPosition;
+    }
+
     private POJOUserEntry newEntry(String username, String password) {
         POJOUserEntry entry = new POJOUserEntry();
         entry.username = username;
         entry.password = password;
         entry.gameType = SLConfig.server.defaultGameType;
+        entry.posX = SLConstants.defaultPosition.getX();
+        entry.posY = SLConstants.defaultPosition.getY();
+        entry.posZ = SLConstants.defaultPosition.getZ();
         return entry;
     }
 
     private static class POJOUserEntry {
         public String password, username;
         public int gameType;
+        double posX, posY, posZ;
```

Comment:
Use `Position` here instead of `posX/Y/Z`?

Reply:
I need immutable

Reply:
I'm not sure if Gson can deal with it correctly